HLTrigger/HLTfilters/plugins/HLTDoubletDZ.cc

```diff
@@ -310,348 +342,271 @@ HLTDoubletDZ<T1, T2>::getCollections(edm::Event& iEvent,
-    return false;
-
-  if (reco::deltaR2(r1->phEta(), r1->phPhi(), r2->phEta(), r2->phPhi()) < minDR_ * minDR_)
-    return false;
-
-  return true;
+bool HLTDoubletDZ<l1t::TrackerMuon, l1t::TrackerMuon>::passCutMinDeltaR(l1t::TrackerMuon const& m1,
+                                                                        l1t::TrackerMuon const& m2) const {
+  return (minDR_ < 0 or reco::deltaR2(m1.phEta(), m1.phPhi(), m2.phEta(), m2.phPhi()) >= minDR2_);
 }
 
-template <>
-bool HLTDoubletDZ<l1t::HPSPFTau, l1t::HPSPFTau>::computeDZ(edm::Event& iEvent,
-                                                           l1t::HPSPFTauRef& r1,
-                                                           l1t::HPSPFTauRef& r2) const {
-  const l1t::HPSPFTau& candidate1(*r1);
-  const l1t::HPSPFTau& candidate2(*r2);
-  if (reco::deltaR(candidate1, candidate2) < minDR_)
-    return false;
-
-  // We don't care about minPixHitsForDZ_ with the L1HPSPFTaus,
-  // especially because the pixel does not participate in the L1T
-  if (std::abs(candidate1.leadChargedPFCand()->pfTrack()->vertex().z() -
-               candidate2.leadChargedPFCand()->pfTrack()->vertex().z()) > maxDZ_)
-    return false;
+namespace {
 
-  return true;
-}
+  double getCandidateVZ(reco::RecoChargedCandidate const& cand, bool& isValidVZ, int const minPixHitsForValidVZ) {
+    if (minPixHitsForValidVZ > 0) {
+      auto const track = cand.track();
+      if (not(track.isNonnull() and track.isAvailable() and
+              track->hitPattern().numberOfValidPixelHits() >= minPixHitsForValidVZ)) {
+        isValidVZ = false;
+        return 0;
+      }
+    }
 
-// ------------ method called to produce the data  ------------
-template <typename T1, typename T2>
-bool HLTDoubletDZ<T1, T2>::hltFilter(edm::Event& iEvent,
-                                     const edm::EventSetup& iSetup,
-                                     trigger::TriggerFilterObjectWithRefs& filterproduct) const {
-  // All HLT filters must create and fill an HLT filter object,
-  // recording any reconstructed physics objects satisfying (or not)
-  // this HLT filter, and place it in the Event.
-  bool accept(false);
+    isValidVZ = true;
+    return cand.vz();
+  }
 
-  std::vector<T1Ref> coll1;
-  std::vector<T2Ref> coll2;
+  double getCandidateVZ(reco::RecoEcalCandidate const& cand,
+                        bool& isValidVZ,
+                        int const minPixHitsForValidVZ,
+                        reco::ElectronCollection const& electrons) {
+    reco::Electron const* elec_ptr = nullptr;
+    for (auto const& elec : electrons) {
+      if (elec.superCluster() == cand.superCluster()) {
+        elec_ptr = &elec;
+      }
+    }
 
-  if (getCollections(iEvent, coll1, coll2, filterproduct)) {
-    int n(0);
-    T1Ref r1;
-    T2Ref r2;
+    if (elec_ptr == nullptr) {
+      // IMPROVE, kept for backward compatibility
+      isValidVZ = true;
+      reco::Electron e1;
+      return e1.vz();
+    }
```

Comment:
Note this ugly choice. It is needed to preserve the current logic of the filter.

In the pre-PR implementation, for example in
https://github.com/cms-sw/cmssw/blob/6535c77febedc693eda635dcfbc765d13a6401bc/HLTrigger/HLTfilters/plugins/HLTDoubletDZ.cc#L340-L354
there is no check that a valid super-cluster match is found; when it isn't (which seems to happen), the pre-PR filter uses `e1.vz()` if `minPixHitsForDZ_ == 0`, and would likely crash if `minPixHitsForDZ_ > 0` (due to an invalid GSF-track ref). Should this logic be left as is, or improved?

Reply:
> Note this ugly choice. It is needed to preserve the current logic of the filter.

By the way, `return 0;` is numerically equivalent to, and likely faster than, [1]. I will update the PR accordingly.

[1]
```cpp
reco::Electron e1;
return e1.vz();
```